### PR TITLE
Data generation testing

### PR DESCRIPTION
**Pull request by Lewis-Kelley**
Wrote tests for most of the code we have written as well as a basic CI pipeline.

### CODE REVIEW

File: data-generation/generate_instructions.py [hunk 1]
```diff
@@ -11,27 +11,28 @@ class NoActionException(Exception):
...
-        return self.side1_letter == other.side1_letter and \
-            self.side2_letter    == other.side2_letter and \
-            self.side1_color     == other.side1_color  and \
-            self.side2_color     == other.side2_color  or  \
-                                                           \
-            self.side1_letter    == other.side2_letter and \
-            self.side2_letter    == other.side1_letter and \
-            self.side1_color     == other.side2_color  and \
-            self.side2_color     == other.side1_color
+        return self.block_id == other.block_id
```
Comment (jcolemang) [inline data-generation/generate_instructions.py:35]: nice

File: data-generation/generate_instructions_test.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+        assert scattered_config.final_blocks == []
+        assert scattered_config.current_blocks == self.scattered_blocks
+
+def test_random_block():
+    indexer = {}
+    indexer['rand_index'] = 0
+    def mock_rand_element(ls):
+        to_return = ls[indexer['rand_index']]
+        indexer['rand_index'] += 1
+        return to_return
```
Comment (jcolemang) [inline data-generation/generate_instructions_test.py]: Why mock `rand_element`?

File: data-generation/generate_instructions_test.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    block = gi.random_block(letters, colors)
+
+    gi.rand_element = orig_rand_element
+    gi.random_position = orig_random_position
+
+    assert block.side1_letter in letters
+    assert block.side2_letter in letters
+    assert block.side1_color in colors
+    assert block.side2_color in colors
+    assert block.position == random_position
```
Comment (jcolemang) [inline data-generation/generate_instructions_test.py]: Why mock the random functions here? The letter and color assertions should hold with or without randomness, and I would think that checking that the position is a tuple containing two numbers would be enough.

**Review by jcolemang — APPROVED** [review]

**Review by kgroble — APPROVED** [review]

**Comment by Lewis-Kelley** [comment]
Merged into neural network project. Closed here.

**Pull request closed without merging**